invenio_rest/csrf.py

```diff
@@ -0,0 +1,262 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2020 CERN.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""CSRF Middleware.
+
+The implementation is highly inspred from Django's initial implementation
+about CSRF protection. For more information you can see here:
+<https://github.com/django/django/blob/master/django/middleware/csrf.py>
+"""
+
+import re
+import secrets
+import string
+
+from flask import Blueprint, abort, current_app, request
+from itsdangerous import BadSignature, URLSafeTimedSerializer
+from six import string_types
+from six.moves.urllib.parse import urlparse
+
+from .errors import RESTCSRFError
+
+REASON_NO_REFERER = "Referer checking failed - no Referer."
+REASON_BAD_REFERER = (
+    "Referer checking failed - %s does not match any trusted origins."
+)
+REASON_NO_CSRF_COOKIE = "CSRF cookie not set."
+REASON_BAD_TOKEN = "CSRF token missing or incorrect."
+REASON_MALFORMED_REFERER = "Referer checking failed - Referer is malformed."
+REASON_INSECURE_REFERER = (
+    "Referer checking failed - Referer is insecure while host is secure."
+)
+
+
+def _get_csrf_serializer():
+    return URLSafeTimedSerializer(
```

Comment:
I would a comment to specify that this token **has** to be signed to avoid any tampering client side in case someone will change implementation in the future.